Update MenuItem and ToolbarItem articles with instructions and examples to enable or disable

brunck opened the pull request:
There has been some confusion when using Xamarin.Forms around how to specify whether a `ToolbarItem` is enabled or disabled. This has changed in recent versions and it is not clear how to do this anymore. This is an attempt to clarify how to do this, including an example.

`ToolbarItem` inherits much of its functionality from `MenuItem` therefore that article was updated as well.

brunck committed:
Update menuitem.md

Note about the IsEnabled property.

brunck committed:
Update toolbaritem.md

Instructions on enabling or disabling a ToolBarItem with example code

brunck committed:
Update toolbaritem.md

edit

brunck committed:
Update toolbaritem.md

edit

brunck committed:
Update menuitem.md

edit

brunck committed:
Update menuitem.md

edit

brunck committed:
Update menuitem.md

add example

brunck committed:
Update menuitem.md

edit

brunck committed:
Update menuitem.md

fix heading capitalization

brunck committed:
Update toolbaritem.md

fix heading capitalization

brunck committed:
Update menuitem.md

fix heading capitalization

brunck committed:
Update toolbaritem.md

fix confusing icon name in example

PRMerger8 commented:
@brunck : Thanks for your contribution! The author(s) have been notified to review your proposed change.

profexorgeek commented:
@brunck thank you for a well-written contribution. Unfortunately it is not correct. The `IsEnabled` property is not read-only as you can see in the API doc here:
https://docs.microsoft.com/en-us/dotnet/api/xamarin.forms.menuitem.isenabled?view=xamarin-forms#Xamarin_Forms_MenuItem_IsEnabled

In the sample project you can test this by altering one of the `MenuItem` initializers like this:
```csharp
MenuItem menuItem2 = new MenuItem
{
   Text = "Delete",
   IsEnabled = false,
};
```

The resulting context menu is greyed-out (example on Android):
![image](https://user-images.githubusercontent.com/711100/70254086-bc702280-1741-11ea-8894-fc1a312ad713.png)

It is possible if you are manipulating underlying properties in an `ICommand` that you could be causing unexpected interactions between the property and the command. However, if you test the `IsEnabled` property in isolation you should find that it is settable and capable of disabling the menu item.

brunck commented:
@profexorgeek Yep, sorry for that. What I intended to do was indicate that the BindableProperty `MenuItem.IsEnabledPropertyKey` is read-only. Obviously got confused.

There is an existing problem that binding to `IsEnabled` in order to toggle the enabled/disabled state of a ToolbarItem does not work. In fact, it acts like the property is ignored altogether as far as data-binding goes. This does not work on both iOS and Android.

The way I describe is how to get that toggling to work. Using the `Command` instead of the `IsEnabled` property to toggle it works (most of the time, there is an existing issue that it doesn't work on Shell apps in Android). That's really what I meant to do here, and that way isn't documented as far as I can tell. 

The data-binding to `IsEnabled` problem doesn't seem like something that will be fixed anytime soon, and it's not clear if it is something to be fixed at all, so I wanted to have some instruction on how to do it the alternate way.

I'll go ahead and update this to reflect that and see if you find it useful in the new form.

profexorgeek commented:
@brunck After reading your comment about IsEnabled and Shell bugs I checked to see if we had any issues related to this and found the issue you filed against Xamarin.Forms:
https://github.com/xamarin/Xamarin.Forms/issues/8741

Your issue has already been reviewed by @PureWeen, tagged and included in vNext (4.4.0). If your proposed documentation is a workaround for an issue that's filed and triaged we should probably wait for the fix!

brunck commented:
@profexorgeek actually that issue was discovered by using the CanExecute approach that I wanted to describe in this PR. It's not an approach that is going to be removed, and I wanted to have some documentation around how to do it that way. 8741 is to have that approach work with Android using Shell. Essentially, this PR describes (once I clean it up) an approach that should be working, and 8741 just ensures that it works as expected on Android.

Issue Xamarin.Forms [5386](https://github.com/xamarin/Xamarin.Forms/issues/5386) addresses the problem with data-binding to IsEnabled on a ToolbarItem. The approach I'm talking about doesn't directly bind to IsEnabled but rather uses a different way (not a workaround) using binding to Command to do the same thing.

I imagine that any fixes to the issues mentioned here wouldn't change the info I want to provide here, so I'd like to go ahead with this.

profexorgeek commented:
Okay sounds good, I just wanted to save you the work if it was a workaround solution!

brunck committed:
remove references to IsEnabled being read-only

PRMerger20 commented:
@brunck : Thanks for your contribution! The author(s) have been notified to review your proposed change.

brunck committed:
Update toolbaritem.md

PRMerger7 commented:
@brunck : Thanks for your contribution! The author(s) have been notified to review your proposed change.

brunck commented:
@profexorgeek I've made some changes.

brunck committed:
Update menuitem.md

PRMerger6 commented:
@brunck : Thanks for your contribution! The author(s) have been notified to review your proposed change.

brunck committed:
Update toolbaritem.md

PRMerger7 commented:
@brunck : Thanks for your contribution! The author(s) have been notified to review your proposed change.

brunck commented:
@profexorgeek bump

profexorgeek commented:
Hey @brunk,
Sorry I missed this! I'll try to take a look in the next couple days.

brunck commented:
@profexorgeek bump again 😈

profexorgeek commented:
Hi @brunck, Sorry for my delay in digging into this. The issue, which I have discussed with our docs team, is that this is bringing to mind some other gotchas/conversations around canexecute and IsEnabled and we want to make sure that this edit is accurate with those conversations!

@davidbritch and I discussed this morning and I'm reassigning this to him as he has a bit more context.

Thanks for keeping this top-of-mind!

davidbritch commented:
Hi @brunck 

Thanks for bringing this to our attention. After investigation, I agree that you're correct. So I will accept this PR, after making some edits to it to bring it inline with our style guide (later today). I'm also going to try and get some movement from the engineers on [the bug](https://github.com/xamarin/Xamarin.Forms/issues/5386).

davidbritch committed:
Brought text inline with our style guide.

PRMerger18 commented:
@davidbritch : Thanks for your contribution! The author(s) have been notified to review your proposed change.

davidbritch committed:
Linked back to MenuItem doc

Rather than duplicate the similar text in the MenuItem doc, I've linked to it instead.

PRMerger19 commented:
@davidbritch : Thanks for your contribution! The author(s) have been notified to review your proposed change.

davidbritch reviewed (approved):
Thanks for the PR @brunck and apologies for the delay in dealing with it.

davidbritch commented:
#sign-off
